nixos/modules/security/krb5.nix

```diff
@@ -0,0 +1,153 @@
+      then "false"
+      else if isInt atom
+      then toString atom
+      else if isString atom
+      then atom
+      else throw "unreachable";
+  in "${name} = ${v}";
+
+  mkRemovedOptionModule' = name: reason: mkRemovedOptionModule ["krb5" name] reason;
+  mkRemovedOptionModuleCfg = name: mkRemovedOptionModule' name ''
+    The option `krb5.${name}' has been removed. Use
+    `security.krb5.settings.${name}' for structured configuration or
+    `security.krb5.extraConfig' for plain-text configuration.
+  '';
+  mkMovedOptionModule = name: mkRemovedOptionModule' name ''
+    The option `krb5.${name}' has been moved to `security.krb5.${name}'.
+  '';
+
+  cfg = config.security.krb5;
+in {
+  imports = [
+    (mkRemovedOptionModuleCfg "libdefaults")
+    (mkRemovedOptionModuleCfg "realms")
+    (mkRemovedOptionModuleCfg "domain_realm")
+    (mkRemovedOptionModuleCfg "capaths")
+    (mkRemovedOptionModuleCfg "appdefaults")
+    (mkRemovedOptionModuleCfg "plugins")
+    (mkRemovedOptionModule' "config" ''
+      The option `krb5.config' has been removed. Use `security.krb5.settings'
+      for structured configuration or `security.krb5.extraConfig' for plain-text
+      configuration.
+    '')
+    (mkMovedOptionModule "extraConfig")
+    (mkMovedOptionModule "kerberos")
+  ];
+
+  options = {
+    security.krb5 = {
+      enable = mkEnableOption (mdDoc "building krb5.conf, configuration file for Kerberos V");
+
+      kerberos = mkOption {
+        type = types.package;
+        default = pkgs.krb5;
+        defaultText = literalExpression "pkgs.krb5";
+        example = literalExpression "pkgs.heimdal";
+        description = lib.mdDoc ''
+          The Kerberos implementation that will be present in
+          `environment.systemPackages` after enabling this
+          service.
+        '';
+      };
+
+      settings = mkOption {
+        default = {};
+        description = mdDoc ''
+          Structured contents of the `krb5.conf` file. See `krb5.conf(5)`
+          for details about configuration.
+        '';
+        example = literalExpression ''
+          {
+            libdefaults = {
+              default_realm = "ATHENA.MIT.EDU";
+            };
+
+            realms = {
+              "ATHENA.MIT.EDU" = {
+                admin_server = "athena.mit.edu";
+                kdc = [
+                  "athena01.mit.edu"
+                  "athena02.mit.edu"
+                ];
+              };
+            };
+
+            domain_realm = {
+              "mit.edu" = "ATHENA.MIT.EDU";
+            };
+
+            logging = {
+              kdc = "SYSLOG:NOTICE";
+              admin_server = "SYSLOG:NOTICE";
+              default = "SYSLOG:NOTICE";
+            };
+          }
+        '';
+        type = toplevel;
+      };
+
+      extraConfig = mkOption {
+        type = lines;
+        default = "";
+        example = ''
+          includedir /etc/krb5.conf.d
+        '';
+        description = mdDoc ''
```

Comment:
Just a suggestion: I think it would be nice if the description had a note that specified that the contents of this setting will be put at the top of the file. Options named `extraConfig` tends to be put at the bottom in other modules.

Reply:
It should not put it at the top of the file, unless I've used mkOrder wrong. Just setting `extraConfig = "blah";` should put it at the bottom of the file, unless you use mkBefore. I'd have to check though.

Reply:
Makes sense, I just confused myself seeing the `mkBefore`. Maybe you could add a comment there below the existing one, that describes the intention rather than the priority details? I don't think it's trivial to figure out at a first glance.